share/spack/win-compiler-wrapper/cl.cxx

```diff
@@ -0,0 +1,406 @@
+    this->spackCommand = spackenv.SpackFC;
+}
+
+
+std::unique_ptr<ToolChainInvocation> ToolChainFactory::ParseToolChain(int argc, char const* const * argv) {
+    std::string command(argv[0]);
+    char const* const* cli(++argv);
+    stripPathandExe(command);
+    try {
+        std::string language = SupportedTools.at(command);
+        std::unique_ptr<ToolChainInvocation> Tool;
+        if(language == "c/c++")
+        {
+            Tool = std::unique_ptr<ClInvocation>(new ClInvocation(command, cli));
+        }
+        else if(language == "intel fortran")
+        {
+            Tool = std::unique_ptr<IntelFortranInvocation>(new IntelFortranInvocation(command, cli));
+        }
+        else{
+            // If it's not c/c++ or fortran we're linking
+            Tool = std::unique_ptr<LdInvocation>(new LdInvocation(command, cli));
+        }
+        return Tool;
+    } catch (std::out_of_range& e) {
+        throw SpackUnknownCompilerException(command);
+    }
+}
+
+void ToolChainFactory::stripPathandExe(std::string &command) {
+    stripPath(command);
+    stripExe(command);
+};
+void ToolChainFactory::stripExe(std::string &command) {
+    std::string::size_type loc = command.rfind(".exe");
+    if (std::string::npos != loc)
+        command.erase(loc);
+}
+
+void ToolChainFactory::stripPath(std::string &command) {
+    command.erase(0, command.find_last_of("\\/") + 1);
+}
+
+const std::map<std::string, std::string> ToolChainFactory::SupportedTools{
+                                                        {"cl", "c/c++"},
+                                                        {"link", "link"},
+                                                        {"ifort", "intel fortran"},
+                                                        {"ifx", "intel fortran"}
+                                                    };
+
+
+void validate_spack_env()
+{
+    std::vector<std::string> SpackEnv{
+"SPACK_ENV_PATH"
+"SPACK_DEBUG_LOG_DIR"
+"SPACK_DEBUG_LOG_ID"
+"SPACK_COMPILER_SPEC"
+"SPACK_CC_RPATH_ARG"
+"SPACK_CXX_RPATH_ARG"
+"SPACK_F77_RPATH_ARG"
+"SPACK_FC_RPATH_ARG"
+"SPACK_LINKER_ARG"
+"SPACK_SHORT_SPEC"
+"SPACK_SYSTEM_DIRS"};
+    for(auto &var: SpackEnv)
+        if(!getenv(var.c_str())){
+            throw SpackCompilerContextException(var);
+        }
+}
+
+void die(std::string & cli ) {
+    throw SpackCompilerException(cli);
+}
+
+SpackList split(std::string s, std::string delim) {
+    size_t pos_start = 0, pos_end;
+    size_t delim_len = delim.length();
+    std::string token;
+    SpackList res;
+
+    while ((pos_end = s.find(delim, pos_start)) != std::string::npos) {
+        size_t token_len = pos_end - pos_start;
+        token = s.substr(pos_start, token_len);
+        pos_start = pos_end + delim_len;
+        res.push_back(token);
+    }
+    res.push_back(s.substr(pos_start));
+    return res;
+}
+
+std::string getSpackEnv(const char * env) {
+    auto envVal = getenv(env);
```

Comment:
Can you avoid using `auto` in this case? I was trying to reason about whether all possible types for `envVal` would make sense when evaluated in the ternary, but I assume that `getenv` returns `const char*` in all cases (so I'd be curious what the handling is for `""` or `null`.

Since the return type is `std::string`, I'm also unclear how this can return `envVal`

Reply:
Sure re: `auto`. The handling is to return an empty string. 
Returning a char buffer as a string is legal c++. The string constructor creates a deep copy of the char buffer.